NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  This file is part of the NineAnimator project.
+//
+//  Copyright © 2018-2020 Marcus Zhou. All rights reserved.
+//
+//  NineAnimator is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  NineAnimator is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with NineAnimator.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+import Foundation
+import SwiftSoup
+
+extension NASourceAnimeHub {
+    static let knownServers = [
+        "fserver": "Fserver",
+        "fdserver": "FDserver",
+        //"xserver": "Xserver", Seems like this no episodes support this server
+        "oserver": "Oserver",
+        "mpserver": "MPserver",
+        "yuserver": "YUserver"
+        //"hserver": "Hserver" Excluding until we update HydraX parser
+    ]
+
+    static let urlRegex = try! NSRegularExpression(pattern: #"((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=\+\$,\w]+@)?[A-Za-z0-9.-]+|(?:www.|[-;:&=\+\$,\w]+@)[A-Za-z0-9.-]+)((?:\/[\+~%\/.\w-_]*)?\??(?:[-\+=&;%@.\w_]*)#?(?:[\w]*))?)"#, options: .caseInsensitive)
+    
+    func episode(from link: EpisodeLink, with anime: Anime) -> NineAnimatorPromise<Episode> {
+        NineAnimatorPromise<String>.firstly {
+            // Extract Episode ID from EpisodeLink URL
+            let episodeURLParams = try URL(string: link.identifier, relativeTo: self.endpointURL).tryUnwrap()
+                .query
+                .tryUnwrap()
+            let episodeID = try formDecode(episodeURLParams).value(at: "ep", type: String.self)
+            return episodeID
+        } .thenPromise {
+            episodeID in
+            // Request api for episode iframe
+            self.requestManager.request(
+                "ajax/anime/load_episodes_v2",
+                handling: .ajax,
+                query: ["s": link.server],
+                parameters: ["episode_id": episodeID]
+            ).responseData.then {
```

Comment:
If you are using `thenPromise` for this block, promote the next `then` statement to the parent promise.

```Swift
// Deprecated

/* First Promise */ .thenPromise {
    return /* Second Promise */.then {
        return /* Process second promise results */
    }
}

// Recommended

/* First Promise */ .thenPromise {
    return /* Second Promise */
} .then {
    return /* Process second promise results */
}
```